File: ios/VulnerabilitiesDetector.swift
struct VulnerabilitiesReport: Encodable {
    var jailbreak: JailbreakReport?
    var debugger: DebuggerReport?
    var emulator: EmulatorReport?
    
    var isSecure: Bool {
        return !(jailbreak?.value ?? false || debugger?.value ?? false)
    }
    
    enum CodingKeys: String, CodingKey {
        case isSecure
        case jailbreak
        case debugger
        case emulator
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(isSecure, forKey: .isSecure)

        if let jailbreak = jailbreak {
            try container.encode(jailbreak, forKey: .jailbreak)
        }

        if let debugger = debugger {
            try container.encode(debugger, forKey: .debugger)
        }
        
        if let emulator = emulator {
            try container.encode(emulator, forKey: .emulator)
        }
    }
}

class VulnerabilitiesDetector: DebuggerDetectorDelegate {
    
    private let debuggerDetector = DebuggerDetector()
    private let jailbreaksDetector = JailbreaksDetector()
    private let emulatorDetector = EmulatorDetector()
    
    private var report: VulnerabilitiesReport?
    
    var delegate: VulnerabilitiesDetectorDelegate?
    
    func debuggerWasAttached(report: DebuggerReport) {
        if self.report == nil {
            self.report = VulnerabilitiesReport(debugger: report)
        } else {
            self.report?.debugger = report
        }
        delegate?.onChange(report: self.report!)
    }
    
    func runDetection() -> VulnerabilitiesReport {
        
        debuggerDetector.delegate = self
        
        let jailbrakeReport = jailbreaksDetector.runDetection()
        let debuggerReport = debuggerDetector.runDetection()
        let emulatorReport = emulatorDetector.runDetection()
        
        report = VulnerabilitiesReport(jailbreak: jailbrakeReport, debugger: debuggerReport, emulator: emulatorReport)
        
        debuggerDetector.startMonitoring()
        
        return report!
    }
}

protocol VulnerabilitiesDetectorDelegate {
    func onChange(report: VulnerabilitiesReport) -> Void
}


extension Encodable {
    func toDictionary() throws -> [String: Any] {
        let data = try JSONEncoder().encode(self)
        let jsonObject = try JSONSerialization.jsonObject(with: data, options: [])
        guard let dict = jsonObject as? [String: Any] else {
            throw NSError(domain: "toDictionary", code: 0, userInfo: [
                NSLocalizedDescriptionKey: "Failed to convert to dictionary"
            ])
        }
        return dict
    }
}

File: android/src/main/java/expo/modules/securitychecker/VulnerabilitiesDetector.kt
package expo.modules.securitychecker

import android.util.Log
import org.json.JSONObject

data class VulnerabilitiesReport(
    val root: RootReport? = null,
    val debugger: DebuggerReport? = null,
    val emulator: EmulatorReport? = null
) {
    val isSecure: Boolean
        get() = !(root?.value == true || debugger?.value == true || emulator?.value == true)
}

fun interface VulnerabilitiesDetectorDelegate {
    fun onChange(report: VulnerabilitiesReport)
}

class VulnerabilitiesDetector(
    private val debuggerDetector: DebuggerDetector = DebuggerDetector(),
    private val rootDetector: RootDetector = RootDetector(),
    private val emulatorDetector: EmulatorDetector = EmulatorDetector()
) : DebuggerDetectorDelegate {

    private var report: VulnerabilitiesReport? = null
    var delegate: VulnerabilitiesDetectorDelegate? = null

    override fun debuggerWasAttached(report: DebuggerReport) {
        if (this.report == null) {
            this.report = VulnerabilitiesReport(debugger = report)
        } else {
            this.report = this.report?.copy(debugger = report)
        }
        delegate?.onChange(this.report!!)
    }

    fun runDetection(): VulnerabilitiesReport {
        debuggerDetector.delegate = this

        val rootReport = rootDetector.runDetection()
        val debuggerReport = debuggerDetector.runDetection()
        val emulatorReport = emulatorDetector.runDetection()

        val result = VulnerabilitiesReport(
            root = rootReport,
            debugger = debuggerReport,
            emulator = emulatorReport
        )

        Log.d("SecurityChecker", "Report = ${result.toJSMap()}" )

        this.report = result

        debuggerDetector.startMonitoring()

        return result
    }
}

fun VulnerabilitiesReport.toJSMap(): Map<String, Any?> {
    return mapOf(
        "isSecure" to isSecure,
        "root" to root?.toJSMap(),
        "debugger" to debugger?.toJSMap(),
        "emulator" to emulator?.toJSMap()
    )
}

fun JSONObject.toMap(): Map<String, Any?> {
    val map = mutableMapOf<String, Any?>()
    val keys = keys()
    while (keys.hasNext()) {
        val key = keys.next()
        map[key] = this.opt(key)
    }
    return map
}